Core/src/org/sleuthkit/autopsy/datamodel/utils/BackgroundTaskRunner.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Autopsy Forensic Browser
+ *
+ * Copyright 2020 Basis Technology Corp.
+ * Contact: carrier <at> sleuthkit <dot> org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.sleuthkit.autopsy.datamodel.utils;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+import java.lang.ref.WeakReference;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.logging.Level;
+import org.openide.nodes.AbstractNode;
+import org.openide.util.WeakListeners;
+import org.sleuthkit.autopsy.coreutils.Logger;
+import org.sleuthkit.autopsy.datamodel.AbstractContentNode;
+
+/**
+ * A utility that allows AbstractNode subclasses to execute background tasks in
+ * threads in a thread pool. An AbstractNode subclass client needs to provide an
+ * implementation of the NodeTask interface that does the background task and
+ * returns the task result in the form of a PropertyChangeEvent. It also needs
+ * to provide an implementation of a PropertyChangeListener to handle the
+ * PropertyChangeEvent returned by the NodeTask. The utility uses weak
+ * references to the AbstractNode and the PropertyChangeListener to safely
+ * return results to the node, if it has not been destroyed by the NetBeans
+ * framework at the time the task is completed.
+ */
+public final class BackgroundTaskRunner {
+
+    private static final Logger logger = Logger.getLogger(AbstractContentNode.class.getName());
+    private static final Integer THREAD_POOL_SIZE = 10;
+    private static final ExecutorService executor = Executors.newFixedThreadPool(THREAD_POOL_SIZE, new ThreadFactoryBuilder().setNameFormat("node-background-task-%d").build());
+
+    /**
+     * Implementations of this interface do a task in a background thread
+     * supplied by this utility.
+     */
+    @FunctionalInterface
+    public interface NodeTask {
+
+        /**
+         * Performs a task in a thread supplied by this utility.
+         *
+         * @param future The future of the actual background task executing this
```

Comment:
Future.cancel() will set the interrupt flag on the underlying thread. What advantages are there of passing the future instance through instead of checking Thread.currentThread().isInterrupted()?

Reply:
This impl invites deadlock. Future.get() will block one of the Threads indefinitely.

Reply:
See reply below.